json_rpc/src/request/params.rs

```diff
@@ -0,0 +1,245 @@
+use std::fmt::{self, Display, Formatter};
+
+use serde_json::{Map, Value};
+
+use super::ErrorOrRejection;
+use crate::error::{Error, ReservedErrorCode};
+
+/// The "params" field of a JSON-RPC request.
+///
+/// As per [the JSON-RPC specification](https://www.jsonrpc.org/specification#parameter_structures),
+/// if present these must be a JSON Array or Object.
+///
+/// `Params` is effectively a restricted [`serde_json::Value`], and can be converted to a `Value`
+/// using `Value::from()` if required.
+#[derive(Clone, Eq, PartialEq, Debug)]
+pub enum Params {
+    /// Represents a JSON Array.
+    Array(Vec<Value>),
+    /// Represents a JSON Object.
+    Object(Map<String, Value>),
+}
+
+impl Params {
+    pub(super) fn try_from(request_id: &Value, params: Value) -> Result<Self, ErrorOrRejection> {
+        let err_invalid_request = |additional_info: &str| {
+            let error = Error::new(ReservedErrorCode::InvalidRequest, additional_info);
+            Err(ErrorOrRejection::Error {
+                id: request_id.clone(),
+                error,
+            })
+        };
```

Comment:
nit: I'd have this as a private function outside `impl Params`.

Reply:
I think this is fine - it's a tiny function inside another small function, and it avoids having to pass in the `id` at each callsite.  Not a big deal either way, but I'd rather leave as is.